### ISSUE TITLE
Make optional config, fail for parse errors instead of returning None

### ISSUE THREAD

**Issue by afsalthaj** [zio/zio-config#127]
Example:

```scala
object ConfigSpec
    extends DefaultRunnableSpec(
      suite("config suite")(
        testM("optional Int example") {

          case class ConfigInternal private (
            domain: String,
            maybePort: Option[Int]
          )
          val envVars: Map[String, String] = Map(
            "DOMAIN" -> "dom1",
            "PORT"   -> "NOT_AN_INT"
          )
          val configDescriptor: ConfigDescriptor[String, String, ConfigInternal] =
            (string("DOMAIN") |@| int(
              "PORT"
            ).optional)(ConfigInternal.apply, ConfigInternal.unapply)

          val configM: IO[ReadErrors[Vector[String], String], config.Config[ConfigInternal]] =
            zio.config.Config.fromMap(envVars, configDescriptor)

          for {
            c        <- configM
            internal <- c.config.config
          } yield assert(internal, equalTo(ConfigInternal("dom1", None)))
          // succeeds but we would expect Optional[Int] field to fail when given data "NOT_AN_INT"
        }
      )
    )
```

**Comment by drmarjanovic** [comment]
@afsalthaj, I would like to take this issue if you agree?

**Comment by afsalthaj** [comment]
@drmarjanovic  I think I missed out your message. Sorry for late response.
Ofcourse, please take it up .Thanks.

**Comment by afsalthaj** [comment]
Feel free to ping me in discord or here anytime.

**Comment by afsalthaj** [comment]
This is now being tracked here, with errors being typed in a better way
https://github.com/zio/zio-config/blob/fix_list_bug/core/src/main/scala/zio/config/ReadFunctions.scala#L115

**Comment by afsalthaj** [comment]
A few behaviours such as;
Optional: 
If there is a parse or unknown error occurred for a key, then its `optional` node will still return error.

OrElseEither: Either[A, B]
If there is any error, parse error or missing value when trying to get, then OrElseEither node tries if it can produce B. For unknown errors, it will avoid trying to get B

`Example: string("key1").orElseEither(int("key2"))`

OrElse: A (it is in fact, Either[A, A])
OrElse node works, if the first key is missing. All the other errors will result in not trying the second key.

`Ex: string("key1").orElse(string("key2)`

**Comment by afsalthaj** [comment]
Fixed in #236
